File: backend/app/schemas/room_availability.py
# backend/app/schemas/room_availability.py

from pydantic import BaseModel, field_validator, Field
from typing import Optional, List, Dict, Any
from datetime import datetime, date as Date  # ← CORRIGIDO: Alias para evitar conflito
from decimal import Decimal


class RoomAvailabilityBase(BaseModel):
    """Schema base para RoomAvailability"""
    room_id: int = Field(..., description="ID do quarto")
    date: Date = Field(..., description="Data da disponibilidade")  # ← CORRIGIDO: Usando alias
    
    # Status de disponibilidade
    is_available: bool = Field(default=True, description="Disponível para reserva")
    is_blocked: bool = Field(default=False, description="Bloqueado manualmente")
    is_out_of_order: bool = Field(default=False, description="Fora de funcionamento")
    is_maintenance: bool = Field(default=False, description="Em manutenção")
    
    # Preço e restrições
    rate_override: Optional[Decimal] = Field(None, ge=0, description="Preço específico para esta data")
    min_stay: int = Field(default=1, ge=1, le=30, description="Estadia mínima")
    max_stay: Optional[int] = Field(None, ge=1, le=365, description="Estadia máxima")
    
    # Restrições de check-in/check-out
    closed_to_arrival: bool = Field(default=False, description="Fechado para chegada")
    closed_to_departure: bool = Field(default=False, description="Fechado para saída")
    
    # Informações adicionais
    reason: Optional[str] = Field(None, max_length=100, description="Motivo do bloqueio")
    notes: Optional[str] = Field(None, max_length=1000, description="Observações")
    metadata_json: Optional[str] = Field(None, max_length=1000, description="Metadados JSON")


class RoomAvailabilityCreate(RoomAvailabilityBase):
    """Schema para criação de RoomAvailability"""
    
    @field_validator('date')
    @classmethod
    def validate_date(cls, v):
        from datetime import date as today_date  # Import local para evitar conflito
        if v < today_date.today():
            raise ValueError('Data não pode ser no passado')
        return v
    
    # REMOVIDO: validator problemático do max_stay
    # Validação será feita no service layer ao invés do schema


class RoomAvailabilityUpdate(BaseModel):
    """Schema para atualização de RoomAvailability"""
    is_available: Optional[bool] = None
    is_blocked: Optional[bool] = None
    is_out_of_order: Optional[bool] = None
    is_maintenance: Optional[bool] = None
    
    rate_override: Optional[Decimal] = Field(None, ge=0)
    min_stay: Optional[int] = Field(None, ge=1, le=30)
    max_stay: Optional[int] = Field(None, ge=1, le=365)
    
    closed_to_arrival: Optional[bool] = None
    closed_to_departure: Optional[bool] = None
    
    reason: Optional[str] = Field(None, max_length=100)
    notes: Optional[str] = Field(None, max_length=1000)
    metadata_json: Optional[str] = Field(None, max_length=1000)
    
    # REMOVIDO: validator problemático que causava recursão
    # Validação será feita no service layer


class RoomAvailabilityResponse(RoomAvailabilityBase):
    """Schema para resposta de RoomAvailability"""
    id: int
    tenant_id: int
    
    # Status de reserva
    is_reserved: bool
    reservation_id: Optional[int]
    
    created_at: datetime
    updated_at: datetime
    is_active: bool
    
    # Campos computados
    status: Optional[str] = None
    is_bookable: Optional[bool] = None
    
    # Dados relacionados
    room_number: Optional[str] = None
    room_name: Optional[str] = None
    room_type_name: Optional[str] = None
    property_name: Optional[str] = None
    
    class Config:
        from_attributes = True


class RoomAvailabilityListResponse(BaseModel):
    """Schema para lista de disponibilidades"""
    availabilities: List[RoomAvailabilityResponse]
    total: int
    page: int
    pages: int
    per_page: int


class RoomAvailabilityFilters(BaseModel):
    """Schema para filtros de busca"""
    room_id: Optional[int] = None
    property_id: Optional[int] = None
    room_type_id: Optional[int] = None
    
    date_from: Optional[Date] = None
    date_to: Optional[Date] = None
    
    is_available: Optional[bool] = None
    is_blocked: Optional[bool] = None
    is_out_of_order: Optional[bool] = None
    is_maintenance: Optional[bool] = None
    is_reserved: Optional[bool] = None
    is_bookable: Optional[bool] = None
    
    closed_to_arrival: Optional[bool] = None
    closed_to_departure: Optional[bool] = None
    
    has_rate_override: Optional[bool] = None
    min_rate: Optional[Decimal] = Field(None, ge=0)
    max_rate: Optional[Decimal] = Field(None, ge=0)
    
    search: Optional[str] = None


# Schemas para operações específicas
class BulkAvailabilityUpdate(BaseModel):
    """Schema para atualização em massa"""
    room_ids: List[int] = Field(..., min_length=1, max_length=100)
    date_from: Date = Field(..., description="Data inicial")
    date_to: Date = Field(..., description="Data final")
    
    # Campos para atualizar
    is_available: Optional[bool] = None
    is_blocked: Optional[bool] = None
    is_out_of_order: Optional[bool] = None
    is_maintenance: Optional[bool] = None
    
    rate_override: Optional[Decimal] = Field(None, ge=0)
    min_stay: Optional[int] = Field(None, ge=1, le=30)
    max_stay: Optional[int] = Field(None, ge=1, le=365)
    
    closed_to_arrival: Optional[bool] = None
    closed_to_departure: Optional[bool] = None
    
    reason: Optional[str] = Field(None, max_length=100)
    notes: Optional[str] = Field(None, max_length=1000)
    
    # REMOVIDO: validator problemático que causava recursão
    # Validação será feita no service layer


class CalendarAvailabilityRequest(BaseModel):
    """Schema para buscar disponibilidade do calendário"""
    room_ids: Optional[List[int]] = None
    property_id: Optional[int] = None
    room_type_id: Optional[int] = None
    date_from: Date = Field(..., description="Data inicial")
    date_to: Date = Field(..., description="Data final")
    include_reserved: bool = Field(default=True, description="Incluir quartos reservados")


class CalendarAvailabilityResponse(BaseModel):
    """Schema para resposta do calendário"""
    date: Date
    availabilities: List[RoomAvailabilityResponse]
    summary: Dict[str, int]  # available, blocked, reserved, etc.


class AvailabilityStatsResponse(BaseModel):
    """Schema para estatísticas de disponibilidade"""
    total_rooms: int = 0
    available_rooms: int = 0
    blocked_rooms: int = 0
    reserved_rooms: int = 0
    maintenance_rooms: int = 0
    out_of_order_rooms: int = 0
    occupancy_rate: float = 0.0
    availability_rate: float = 0.0

File: frontend/src/types/room-availability.ts
// frontend/src/types/room-availability.ts

export interface RoomAvailabilityResponse {
  id: number;
  room_id: number;
  date: string;
  tenant_id: number;
  
  // Status de disponibilidade
  is_available: boolean;
  is_blocked: boolean;
  is_out_of_order: boolean;
  is_maintenance: boolean;
  is_reserved: boolean;
  reservation_id?: number;
  
  // Preços e restrições
  rate_override?: number;
  min_stay: number;
  max_stay?: number;
  
  // Restrições de check-in/check-out
  closed_to_arrival: boolean;
  closed_to_departure: boolean;
  
  // Informações adicionais
  reason?: string;
  notes?: string;
  
  // Metadados
  created_at: string;
  updated_at: string;
  is_active: boolean;
  
  // Campos computados
  status?: string;
  is_bookable?: boolean;
  
  // Dados relacionados
  room_number?: string;
  room_name?: string;
  room_type_name?: string;
  property_name?: string;
}

export interface RoomAvailabilityCreate {
  room_id: number;
  date: string;
  
  // Status de disponibilidade
  is_available?: boolean;
  is_blocked?: boolean;
  is_out_of_order?: boolean;
  is_maintenance?: boolean;
  
  // Preços e restrições
  rate_override?: number;
  min_stay?: number;
  max_stay?: number;
  
  // Restrições de check-in/check-out
  closed_to_arrival?: boolean;
  closed_to_departure?: boolean;
  
  // Informações adicionais
  reason?: string;
  notes?: string;
}

export interface RoomAvailabilityUpdate {
  // Status de disponibilidade
  is_available?: boolean;
  is_blocked?: boolean;
  is_out_of_order?: boolean;
  is_maintenance?: boolean;
  
  // Preços e restrições
  rate_override?: number;
  min_stay?: number;
  max_stay?: number;
  
  // Restrições de check-in/check-out
  closed_to_arrival?: boolean;
  closed_to_departure?: boolean;
  
  // Informações adicionais
  reason?: string;
  notes?: string;
}

export interface RoomAvailabilityListResponse {
  availabilities: RoomAvailabilityResponse[];
  total: number;
  page: number;
  pages: number;
  per_page: number;
}

export interface RoomAvailabilityFilters {
  room_id?: number;
  property_id?: number;
  room_type_id?: number;
  
  date_from?: string;
  date_to?: string;
  
  is_available?: boolean;
  is_blocked?: boolean;
  is_out_of_order?: boolean;
  is_maintenance?: boolean;
  is_reserved?: boolean;
  is_bookable?: boolean;
  
  closed_to_arrival?: boolean;
  closed_to_departure?: boolean;
  
  has_rate_override?: boolean;
  min_rate?: number;
  max_rate?: number;
  
  search?: string;
}

export interface BulkAvailabilityUpdate {
  room_ids: number[];
  date_from: string;
  date_to: string;
  
  // Campos para atualizar
  is_available?: boolean;
  is_blocked?: boolean;
  is_out_of_order?: boolean;
  is_maintenance?: boolean;
  
  rate_override?: number;
  min_stay?: number;
  max_stay?: number;
  
  closed_to_arrival?: boolean;
  closed_to_departure?: boolean;
  
  reason?: string;
  notes?: string;
}

export interface CalendarAvailabilityRequest {
  room_ids?: number[];
  property_id?: number;
  room_type_id?: number;
  date_from: string;
  date_to: string;
  include_reserved?: boolean;
}

export interface CalendarAvailabilityResponse {
  date: string;
  availabilities: RoomAvailabilityResponse[];
  summary: Record<string, number>; // available, blocked, reserved, etc.
}

export interface AvailabilityStatsResponse {
  total_rooms: number;
  available_rooms: number;
  blocked_rooms: number;
  reserved_rooms: number;
  maintenance_rooms: number;
  out_of_order_rooms: number;
  occupancy_rate: number;
  availability_rate: number;
}

export interface RoomAvailabilityCheckRequest {
  room_id: number;
  check_in_date: string;
  check_out_date: string;
}

export interface RoomAvailabilityCheckResponse {
  available: boolean;
  conflicts?: string[];
  blocked_dates?: string[];
  reason?: string;
}

// Status colors para UI
export const AVAILABILITY_STATUS_COLORS: Record<string, string> = {
  available: 'bg-green-100 text-green-800 border-green-200',
  blocked: 'bg-red-100 text-red-800 border-red-200',
  reserved: 'bg-blue-100 text-blue-800 border-blue-200',
  maintenance: 'bg-yellow-100 text-yellow-800 border-yellow-200',
  out_of_order: 'bg-gray-100 text-gray-800 border-gray-200',
};

// Status labels para UI
export const AVAILABILITY_STATUS_LABELS: Record<string, string> = {
  available: 'Disponível',
  blocked: 'Bloqueado',
  reserved: 'Reservado',
  maintenance: 'Manutenção',
  out_of_order: 'Fora de Ordem',
};